### PR TITLE
Laboratory work #4, Margarita Velmisova - 22FPL1

### PR DESCRIPTION
**Pull request by velgarita**

### CODE REVIEW

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]
Nothing to review

File: lab_4_fill_words_by_ngrams/main.py [hunk 1]
```diff
@@ -28,6 +30,19 @@ def _tokenize(self, text: str) -> tuple[str, ...]:  # type: ignore
...
+        if not (isinstance(text, str) and text):
+            raise ValueError('Incorrect input')
+
+        tokenized_word = []
+        for word in text.lower().split():
+            if word.isalpha():
+                tokenized_word.append(word)
+            elif word[:-1].isalpha() and word[-1] in {'.', '!', '?'}:
+                tokenized_word.extend([word[:-1], self._end_of_word_token])
+            elif word[-1] in ['.', '!', '?']:
```
Comment (artyomtugaryov) [inline lab_4_fill_words_by_ngrams/main.py:42]: why not set?

Comment (artyomtugaryov) [reply lab_4_fill_words_by_ngrams/main.py:42]: create it once before the loop

File: lab_4_fill_words_by_ngrams/main.py [hunk 2]
```diff
@@ -125,6 +214,10 @@ def get_conversion_generator_type(self, generator_type: int) -> str:  # type: ig
         Returns:
             (str): Name of the generator.
         """
+        generator_types = {0: 'Greedy Generator', 1: 'Top-P Generator', 2: 'Beam Search Generator'}
```
Comment (artyomtugaryov) [inline lab_4_fill_words_by_ngrams/main.py:217]: this map shoud be in self

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_4_fill_words_by_ngrams/main.py (hunk 2 above)
Comment (velgarita) [reply lab_4_fill_words_by_ngrams/main.py:217]: changed it

File: lab_4_fill_words_by_ngrams/main.py (hunk 1 above)
Comment (velgarita) [reply lab_4_fill_words_by_ngrams/main.py:42]: fixed it

File: lab_4_fill_words_by_ngrams/main.py [hunk 3]
```diff
@@ -28,6 +32,25 @@ def _tokenize(self, text: str) -> tuple[str, ...]:  # type: ignore
...
+        if not (isinstance(text, str) and text):
+            raise ValueError("Inappropriate input.")
+
+        for digit in ('.', '!', '?'):
+            text = text.replace(digit, f" {self._end_of_word_token} ")
+
+        tokenized_word = []
+        for word in text.lower().split():
+            if word == self._end_of_word_token or word.isalpha() or word.isspace():
+                tokenized_word.append(word)
```
Comment (artyomtugaryov) [inline lab_4_fill_words_by_ngrams/main.py:44]: use `continue` and remove `else`

File: lab_4_fill_words_by_ngrams/main.py [hunk 4]
```diff
@@ -69,7 +117,7 @@ class TopPGenerator:
     """
 
     def __init__(
-        self, language_model: NGramLanguageModel, word_processor: WordProcessor, p_value: float
+            self, language_model: NGramLanguageModel, word_processor: WordProcessor, p_value: float
```
Comment (artyomtugaryov) [inline lab_4_fill_words_by_ngrams/main.py:120]: revert

File: lab_4_fill_words_by_ngrams/main.py [hunk 5]
```diff
@@ -125,6 +219,9 @@ def get_conversion_generator_type(self, generator_type: int) -> str:  # type: ig
         Returns:
             (str): Name of the generator.
         """
+        generator_types = {self.greedy: 'Greedy Generator', self.top_p: 'Top-P Generator',
```
Comment (artyomtugaryov) [inline lab_4_fill_words_by_ngrams/main.py:222]: why is this map not in `__init__`?

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_4_fill_words_by_ngrams/main.py (hunk 3 above)
Comment (velgarita) [reply lab_4_fill_words_by_ngrams/main.py:44]: done

File: lab_4_fill_words_by_ngrams/main.py (hunk 4 above)
Comment (velgarita) [reply lab_4_fill_words_by_ngrams/main.py:120]: done

File: lab_4_fill_words_by_ngrams/main.py (hunk 5 above)
Comment (velgarita) [reply lab_4_fill_words_by_ngrams/main.py:222]: changed it

**Comment by vitazaytseva** [comment]
Well done! Good job!

**Pull request closed without merging**